### PR TITLE
8269820: C2 PhaseIdealLoop::do_unroll get wrong opaque node

### PR DESCRIPTION
**Pull request by rwestrel**
A main loop loses its pre loop. The Opaque1 node for the zero trip
guard of the main loop is assigned control at a Region through which
an If is split. As a result, the Opaque1 is cloned and the zero trip
guard takes a Phi that merges Opaque1 nodes. One of the branch dies
next and as, a result, the zero trip guard has an Opaque1 as input but
at the wrong CmpI input. The assert fires next.

The fix I propose is that if an Opaque1 node that is part of a zero
trip guard is encountered during split if, rather than split if up or
down, instead, assign it the control of the zero trip guard's
control. This way the pattern of the zero trip guard is unaffected and
split if can proceed. I believe it's safe to assign it a later
control:

- an Opaque1 can't be shared

- the zero trip guard can't be the If that's being split

As Vladimir noted, this bug used to not reproduce with loop strip
mining disabled but now always reproduces because the loop
strip mining nest is always constructed. The reason is that the
main loop in this test is kept alive by the LSM safepoint. If the
LSM loop nest is not constructed, the loop is optimized out. I
filed:

https://bugs.openjdk.org/browse/JDK-8297724

for this issue.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8269820](https://bugs.openjdk.org/browse/JDK-8269820): C2 PhaseIdealLoop::do_unroll get wrong opaque node

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [26a002f5](https://git.openjdk.org/jdk/pull/11391/files/26a002f5f658062112259b4da5b02e09278f3803)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [26a002f5](https://git.openjdk.org/jdk/pull/11391/files/26a002f5f658062112259b4da5b02e09278f3803)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [26a002f5](https://git.openjdk.org/jdk/pull/11391/files/26a002f5f658062112259b4da5b02e09278f3803)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11391/head:pull/11391` \
`$ git checkout pull/11391`

Update a local copy of the PR: \
`$ git checkout pull/11391` \
`$ git pull https://git.openjdk.org/jdk pull/11391/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11391`

View PR using the GUI difftool: \
`$ git pr show -t 11391`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11391.diff">https://git.openjdk.org/jdk/pull/11391.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/split_if.cpp [hunk 1]
```diff
@@ -220,6 +221,29 @@ bool PhaseIdealLoop::split_up( Node *n, Node *blk1, Node *blk2 ) {
...
+      Node* ctrl = cl->skip_predicates();
+      assert(ctrl->in(0)->is_If(), "expect and If from the zero trip guard");
+      assert(ctrl->in(0) != blk1 && ctrl->in(0) != blk2, "shouldn't split a zero trip guard");
+      Node* bol = ctrl->in(0)->in(1);
+      assert(bol->is_Bool(), "bad zero trip guard shape");
+      Node* cmp = bol->in(1);
+      assert(cmp->Opcode() == Op_CmpI, "bad zero trip guard shape");
+      set_ctrl(n, ctrl->in(0)->in(0));
+      set_ctrl(cmp, ctrl->in(0)->in(0));
+      set_ctrl(bol, ctrl->in(0)->in(0));
```
Comment (vnkozlov) [inline src/hotspot/share/opto/split_if.cpp:242]: Why you assign control to `cmp` and `bol` too?

**Review by vnkozlov — COMMENTED** [review]
General question. Will it help (simplify changes) if we add specialized `class OpaqueZeroTripGuardNode : public Opaque1Node` class?

**Comment by rwestrel** [comment]
Thanks for reviewing this.

> General question. Will it help (simplify changes) if we add specialized `class OpaqueZeroTripGuardNode : public Opaque1Node` class?

Good suggestion! Let me give it a try.

File: src/hotspot/share/opto/split_if.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/split_if.cpp:242]: The subgraph is` (Bool (CmpI (Opaque1...))`. So if only the control of `Opaque1` is updated then the `CmpI`/`Bool` could end up with a control that strictly dominates the control of the `Opaque1`. It's quite possible that that wouldn't break anything but wouldn't it be inconsistent and quite ugly?

Comment (vnkozlov) [reply src/hotspot/share/opto/split_if.cpp:242]: I mean we never assign control edge to `Bool` and `Cmp` nodes - they depend only on their inputs. At least I don't know about it.

Comment (rwestrel) [reply src/hotspot/share/opto/split_if.cpp]: `set_ctrl()` doesn't change the control input of the nodes, right? It only updates the current loop opts pass's table of controls and all data nodes are in that table. I'm confused by what could be wrong here.

**Comment by rwestrel** [comment]
> > General question. Will it help (simplify changes) if we add specialized `class OpaqueZeroTripGuardNode : public Opaque1Node` class?
> 
> Good suggestion! Let me give it a try.

I updated the patch with an OpaqueZeroTripGuardNode.

File: src/hotspot/share/opto/split_if.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/split_if.cpp]: You are right, I forgot that this function only update site table. Good.

**Review by vnkozlov — APPROVED** [review]
Looks nice! And Tobias's testing results also looks good so far (only known failures).

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. Testing passed.

**Review by chhagedorn — APPROVED** [review]
That looks reasonable to me. Good idea to introduce a new opaque node type!

**Comment by rwestrel** [comment]
@vnkozlov @TobiHartmann @chhagedorn thanks for the reviews.

**Pull request closed without merging**